torch/csrc/distributed/rpc/init.cpp

```diff
@@ -405,6 +406,42 @@ If the future completes with an error, an exception is thrown.
                   >>> # The inserted callback will print the return value when
                   >>> # receiving the response from "worker1"
                   >>> fut.add_done_callback(callback)
+          )")
+      .def(
+          "set_result",
+          [&](FutureIValue& fut, py::object result) {
+            fut.markCompleted(torch::jit::toIValue(
+                std::move(result),
+                PyObjectType::get()));
+          },
+          py::call_guard<py::gil_scoped_release>(),
+          R"(
+              Set the result for this ``Future``, which will mark this
+              ``Future`` as completed and trigger all attached callbacks. Note
+              that as a ``Future`` cannot be marked completed twice,
+              applications should never call this method on a ``Future``
+              returned by :meth:`~torch.distributed.rpc.rpc_async`, because the
+              RPC system will also set result for that ``Future`` when the
+              response is received. This method should only be called on
```

Comment:
Do we know what error will happen if the user accidentally does this? I feel like we should add a check if possible to ensure that this doesn't happen - is there a way to distinguish between user's future and RPC system future easily?

If not, then for now we can also consider checking `fut.isCompleted()` in set_result, and raise an error to the user indicating they may be trying to use it incorrectly. Although, this wouldn't catch the case where the user calls set_result, and after that, the RPC system completes the future

Reply:
Let me add a test to check the error message. Current code in `future.h` already checks `completed_` in `markCompleted()`. Will make sure the error message is clear.